pkgs/development/libraries/flutter-engine/package.nix

```diff
@@ -0,0 +1,218 @@
+    installPhase = ''
+      mkdir -p $out
+      cp -r -P --no-preserve=mode,ownership $buildroot $out/src
+      cp -r -P --no-preserve=mode,ownership $src $out/src/flutter
+
+      ${concatStringsSep "\n" (attrValues (mapAttrs (name: src: ''
+        if [[ -d ${src} ]]; then
+          mkdir -p $out/${name}
+          cp -r -P --no-preserve=mode,ownership ${src}/* $out/${name}
+          find ${src} -type f -name '.*' | xargs -I {} cp --no-preserve=mode,ownership {} $out/${name}
+          ${optionalString (builtins.hasAttr "rev" src) ''
+            mkdir -p $out/${name}/.git/logs
+            echo "${src.rev}" >$out/${name}/.git/logs/HEAD
+          ''}
+        else
+          cp -P --no-preserve=mode,ownership ${src} $out/${name}
+        fi
+      '') flutter-deps))}
+
+      touch $out/src/third_party/dart/build/config/gclient_args.gni
+      cp $gitrev $out/src/flutter/build/git_revision.py
+
+      for bin in $binaryFixes; do
+        chmod 0755 $out/$bin
+        patchelf --set-interpreter ${stdenv.cc.libc}/lib/${interpreter} $out/$bin
+      done
+
+      for sh in $shellFixes; do
+        chmod 0755 $out/$sh
+        patchShebangs $out/$sh
+      done
+
+      chmod +x $out/src/build/linux/sysroot_ld_path.sh
+
+      sed -i '1 s|^.*$|#!${gnumake}/bin/make -f|' $out/src/third_party/harfbuzz/src/update-unicode-tables.make
+
+      mkdir -p $out/src/build/linux/debian_sid_${toolchainArch}-sysroot
+      tar -xf $toolchain -C $out/src/build/linux/debian_sid_${toolchainArch}-sysroot
+
+      cd $out
+      mkdir -p $out/src/third_party/dart/.dart_tool
+      cp -r --no-preserve=ownership,mode $dartPackageConfig $out/src/third_party/dart/.dart_tool/package_config.json
+
+      python3 $out/src/third_party/dart/tools/generate_package_config.py
+      python3 $out/src/third_party/dart/tools/generate_sdk_version_file.py
+      python3 $out/src/tools/remove_stale_pyc_files.py src/tools
+      python3 $out/src/flutter/tools/pub_get_offline.py
+    '';
+  };
+in stdenvNoCC.mkDerivation rec {
+  pname = "flutter-engine-${runtimeMode}";
+  version = builtins.substring 0 7 args.version;
+  inherit runtimeMode src;
+
+  nativeBuildInputs = [ git python3 pkg-config ninja stdenv.cc.libc ];
+
+  ## FIXME: create a sysroot toolchain for Nix and use it instead
+  ## TODO: generate for other platforms based on "targetPlatform"
+  configurePhase = ''
+    ./src/flutter/tools/gn \
+      --runtime-mode $runtimeMode \
+      --no-build-glfw-shell \
+      --prebuilt-dart-sdk \
+      --embedder-for-target \
+      --no-goma \
+      --out-dir $out/lib/flutter \
+      --target-dir $runtimeMode
+
+    mv $out/lib/flutter/out/$runtimeMode $out/lib/flutter/$runtimeMode
+    rmdir $out/lib/flutter/out
+  '';
+
+  ## TODO: add more builds and possibly better cross-compiling
+  buildPhase = ''
+    cd $out/lib/flutter/$runtimeMode
+    export TERM=dumb
+
+    echo "Building flatc"
+    ninja flatc -j$NIX_BUILD_CORES
```

Comment:
the build system ignores $NIX_BUILD_CORES
this will cause problems on build servers, usually OOM
and random "internal compiler error: Segmentation fault"
this can be hard to fix, but should at least be mentioned in a comment

also, the `flutter-engine-${runtimeMode}` expression should have

```nix
{
  requiredSystemFeatures = [ "big-parallel" ];

  meta = with lib; {
    # This build takes a long time; particularly on slow architectures
    # TODO update times
    # 1 hour on 32x3.6GHz -> maybe 12 hours on 4x2.4GHz
    timeout = 24 * 3600;
  };
```

Reply:
Oh interesting, it takes only about 10 minutes on my system but I also have a Ryzen 5 3600. I'll add an argument to disable `-j$NIX_BUILD_CORES`.

Reply:
I've added an argument to disable parallel building.

Reply:
aah, found the problem

`ninja -j$NIX_BUILD_CORES` works as expected
but each clang uses multiple cores, so the cpu load is higher

so this should be `ninja -j$NIX_BUILD_CORES -l$NIX_BUILD_CORES` to limit jobs and load